src/vs/editor/contrib/suggest/suggestController.ts

```diff
@@ -427,6 +433,32 @@ export class SuggestController implements IEditorContribution {
 
 		// clear only now - after all tasks are done
 		Promise.all(tasks).finally(() => {
+
+			const resourceLoggedEvents = this._loggedEvents.get(model.uri) ?? new Set<string>();
+			if (!this._loggedEvents.has(model.uri)) {
+				this._loggedEvents.set(model.uri, resourceLoggedEvents);
+			}
+
+			const providerId = event.item.provider._source ?? event.item.provider._debugDisplayName ?? 'unknownProvider';
+			if (!resourceLoggedEvents.has(providerId)) {
+				resourceLoggedEvents.add(providerId);
```

Comment:
Suggestion to extract this into a separate function, similar to [this](https://github.com/microsoft/vscode/blob/a704a69394a3fbd241feaa6c0305067f25f866f5/src/vs/editor/contrib/suggest/suggestModel.ts#L5560). Maybe also wrapped in a setTimeout or idleCallback to ensure this doesn't impact editor rendering perf (which seems unlikely)